frontend/apps/web/components/RunTImeline/RunTimeline.tsx

```diff
@@ -0,0 +1,457 @@
+    }
+
+    if (date instanceof Date) {
+      return format(date, 'MM/dd/yyyy HH:mm:ss:SSS');
+    }
+  };
+
+  const formatDate = (date: Date) => format(date, 'MM/dd/yyyy');
+  const formatTime = (date: Date) => format(date, 'HH:mm:ss:SSS');
+
+  const formatTaskDuration = (
+    s: Timestamp | undefined,
+    end: Date | undefined
+  ) => {
+    if (!s || !end) return 'N/A';
+    const start = convertTimestampToDate(s);
+    const duration = intervalToDuration({ start, end });
+    const milliseconds = end.getTime() - start.getTime();
+    const millis = milliseconds % 1000;
+
+    // format the duration string
+    const formattedDuration = formatDuration(duration, {
+      format: ['hours', 'minutes', 'seconds'],
+      delimiter: ', ',
+    });
+
+    if (!formattedDuration) {
+      return `${millis} ms`;
+    }
+
+    return `${formattedDuration}${millis > 0 ? `, ${millis} ms` : ''}`;
+  };
+
+  const { timelineStart, totalDuration, timeLabels } = useMemo(() => {
+    // find earliest start date out of all of the activities
+    let startTime = Infinity;
+    let endTime = -Infinity;
+
+    tasks.map((t) => {
+      startTime = Math.min(
+        startTime,
+        convertTimestampToDate(t.startTime).getTime()
+      );
+
+      const errorDate = getCloseOrErrorOrCancelDate(t);
+      endTime = Math.max(
+        endTime,
+        errorDate.getTime(),
+        convertTimestampToDate(t.closeTime || t.startTime).getTime()
+      );
+    });
+
+    const start = new Date(startTime);
+    const end = new Date(endTime);
+
+    let duration = end.getTime() - start.getTime();
+
+    // add padding, but limit it to a maximum of 100ms on each side so we can see the entire timeline in view in the graph
+    const padding = Math.min(duration * 0.1, 300);
+    const adjustedStart = new Date(start.getTime() - padding);
+    const adjustedEnd = new Date(end.getTime() + padding);
+    const adjustedDuration = adjustedEnd.getTime() - adjustedStart.getTime();
+
+    const labelCount = 5;
+
+    const labels: Date[] = Array.from({ length: labelCount }, (_, i) =>
+      addMilliseconds(adjustedStart, (adjustedDuration * i) / (labelCount - 1))
+    );
+
+    return {
+      timelineStart: adjustedStart,
+      timelineEnd: adjustedEnd,
+      totalDuration: adjustedDuration,
+      timeLabels: labels,
+    };
+  }, [tasks]);
+
+  // calculates where in the timeline axis something should be relative to the total duration
+  // also dictates how far right the timeline goes, reduce if you want the timeline shorter or length otherwise
+  const getPositionPercentage = (time: Date) => {
+    return ((time.getTime() - timelineStart.getTime()) / totalDuration) * 92;
+  };
+
+  // handles getting the activity statuses by remaping the ActivityStatuses since (i think) we don't want to show all of them per activity
+  // prob want to update with actual types instead of using strings here
+  // the event types in the types field are just the stringified Temporal types
```

Comment:
Since they come back as string, you can always created your own type that they will get parsed to since you are already doing the parsing in this method.

Reply:
updated